### PR TITLE
4099 ES6 Refactor lib/Compiler.js

### PR DESCRIPTION
**Pull request by KTruong888**
**What kind of change does this PR introduce?**

ES6 Refactor of lib/Compiler.js module

**Did you add tests for your changes?**

Yes

**Summary**

Refactored the lib/Compiler.js module to use ES6 as per issue: https://github.com/webpack/webpack/issues/4099

**Other Notes**

I'm also trying to refactor the lib/HotModuleReplacementPlugin and I'm getting a feeling that that file was not meant to be touched because the error messages there are odd. Should I just stop work on that module or is that one meant to be ES6-ified too?

Edit: 

Local test suite passed but the checks failed when I pushed. Closed to let others know I am not currently working on this.

edit2:

Tried it again and tests seem to pass now but the test coverage dropped??

edit3:

Added some tests

### CODE REVIEW

File: lib/Compiler.js [hunk 1]
```diff
@@ -2,520 +2,524 @@
...
-Watching.prototype._go = function() {
-	var self = this;
-	self.startTime = Date.now();
-	self.running = true;
-	self.invalid = false;
-	self.compiler.applyPluginsAsync("watch-run", self, function(err) {
-		if(err) return self._done(err);
-		self.compiler.compile(function onCompiled(err, compilation) {
+	_go() {
+		const self = this;
```
Comment (sokra) [inline lib/Compiler.js]: `self` was only an alias for `this`. This is no longer needed with arrow functions

File: lib/Compiler.js [hunk 2]
```diff
@@ -2,520 +2,524 @@
...
+	}
 
-Compiler.Watching = Watching;
-Compiler.prototype.watch = function(watchOptions, handler) {
-	this.fileTimestamps = {};
-	this.contextTimestamps = {};
-	var watching = new Watching(this, watchOptions, handler);
-	return watching;
-};
+	static Watching(compiler, watchOptions, handler) {
```
Comment (sokra) [inline lib/Compiler.js]: That's not the same. 

Use `Compiler.Watching = Watching;` instead.

File: lib/Compiler.js [hunk 3]
```diff
@@ -2,520 +2,524 @@
...
+		this.fileTimestamps = {};
+		this.contextTimestamps = {};
+		const watching = new Watching(this, watchOptions, handler);
+		return watching;
+	}
 
-	self.applyPluginsAsync("before-run", self, function(err) {
-		if(err) return callback(err);
+	run(callback) {
+		const self = this;
```
Comment (sokra) [inline lib/Compiler.js]: `self` was only an alias for `this`. This is no longer needed with arrow functions

File: lib/Compiler.js [hunk 4]
```diff
@@ -2,520 +2,524 @@
...
-	if(!self.recordsInputPath) {
-		self.records = {};
-		return callback();
-	}
-	self.inputFileSystem.stat(self.recordsInputPath, function(err) {
-		// It doesn't exist
-		// We can ignore self.
-		if(err) return callback();
+	readRecords(callback) {
+		const self = this;
```
Comment (sokra) [inline lib/Compiler.js]: `self` was only an alias for `this`. This is no longer needed with arrow functions

File: lib/Compiler.js [hunk 5]
```diff
@@ -2,520 +2,524 @@
...
+			normalModuleFactory: this.createNormalModuleFactory(),
+			contextModuleFactory: this.createContextModuleFactory(),
+			compilationDependencies: []
+		};
+		return params;
+	}
 
-		self.applyPluginsParallel("make", compilation, function(err) {
+	compile(callback) {
+		const self = this;
```
Comment (sokra) [inline lib/Compiler.js]: `self` was only an alias for `this`. This is no longer needed with arrow functions

File: lib/Compiler.js (hunk 1 above)
Comment (KTruong888) [inline lib/Compiler.js]: converted 'self' to 'this' in the _go method. Shifted some normal functions around and converted them to arrow functions.

File: lib/Compiler.js (hunk 2 above)
Comment (KTruong888) [inline lib/Compiler.js]: moved static Watching method to a property near the last lines

File: lib/Compiler.js (hunk 3 above)
Comment (KTruong888) [inline lib/Compiler.js]: converted 'self' to 'this' in the run method. converted old style function declarations into arrow functions and had to shift position of the function block within the method due to hoisting concerns I think.

File: lib/Compiler.js (hunk 4 above)
Comment (KTruong888) [inline lib/Compiler.js]: converted 'self' to 'this' in the readRecords method.

File: lib/Compiler.js (hunk 5 above)
Comment (KTruong888) [inline lib/Compiler.js]: converted references of self to this in the compile method

File: test/Compiler.test.js [hunk 6]
```diff
@@ -298,14 +298,14 @@ describe("Compiler", () => {
...
-			it("should have an accessible static method, Watching", (done) => {
-				const actual = Compiler.Watching(compiler, 1000, err => err);
-				actual.running.should.be.exactly(true);
-				actual.constructor.name.should.be.exactly("Watching");
-				done();
-			});
-		});
+		// describe("static method", () => {
+		// 	it("should have an accessible static method, Watching", (done) => {
+		// 		const actual = Compiler.Watching(compiler, 1000, err => err);
```
Comment (sokra) [inline test/Compiler.test.js]: `const actual = Compiler.Watching(compiler, 1000, err => err);`
`const actual = new Compiler.Watching(compiler, 1000, err => err);`

Comment (sokra) [inline test/Compiler.test.js]: either remove this test or readd it. Don't use commented code.

Comment (KTruong888) [inline test/Compiler.test.js]: Thanks for the review again. Also wondering, would you prefer if I squashed all the commits or just leave it as it is?

File: lib/Compiler.js [hunk 7]
```diff
@@ -2,520 +2,522 @@
...
+					deprecationReported = true;
+				}
+				this.plugin("compilation", (compilation, data) => {
+					data.normalModuleFactory.plugin("parser", parser => {
+						parser.plugin(hook, fn);
+					});
+				});
+			},
+			apply: () => {
+				const args = arguments;
```
Comment (sokra) [inline lib/Compiler.js]: `arguments` is not available in an arrow function.

Replace it with a function.

Spread arguments would be nicer, but they are not support in node 4.

**Comment by sokra** [comment]
Thanks

**Pull request merged**